vshard/replicaset.lua

```diff
@@ -577,13 +577,33 @@ local function replicaset_template_multicallro(prefer_replica, balance)
     local function pick_next_replica(replicaset, now)
         local r
         local master = replicaset.master
+
         if balance then
+            local prefered_zone = replicaset.priority_list[1].zone
             local i = #replicaset.priority_list
             while i > 0 do
                 r = replicaset_balance_replica(replicaset)
                 i = i - 1
                 if r:is_connected() and (not prefer_replica or r ~= master) and
-                   replica_check_backoff(r, now) then
+                    replica_check_backoff(r, now)
+                then
+                    -- Pick a replica according prefered zone (highest priority replica zone) in round-robin manner
+                    if balance == "prefer_zone" and prefered_zone then
+                        local cbi = replicaset.balance_i
+                        local nr = replicaset_balance_replica(replicaset)
+
+                        if prefered_zone and r.zone and r.zone == prefered_zone
+                            and nr.zone and nr.zone ~= prefered_zone
+                            and (not prefer_replica or nr ~= master)
+                        then
+                            -- Reset cursor to the main position if next replica is in different zone.
```

Comment:
Replicas are not sorted by zones. Hence if in the replica list the first one is in zone1, then one in zone2, and others in zone1, your code will skip the tail-replicas always. That doesn't look good.

There are several solutions - don't touch balance_i. Simply skip replicas from other zones manually. Or sort the replicas by zones in the replica list. Or add a key-value dictionary to each replicaset object, where key = zone name, value = replica list in this zone.

I think it might be too complicated to do the last 2 options, so I suggest to just skip not fitting replicas during the iteration. It seems this code solves your problem and is multiple times shorter:
```Diff
@@ -577,33 +577,18 @@ local function replicaset_template_multicallro(prefer_replica, balance)
     local function pick_next_replica(replicaset, now)
         local r
         local master = replicaset.master
-
         if balance then
-            local prefered_zone = replicaset.priority_list[1].zone
+            local prefered_zone
+            if balance == 'prefered_zone' then
+                prefered_zone = replicaset.priority_list[1].zone
+            end
             local i = #replicaset.priority_list
             while i > 0 do
                 r = replicaset_balance_replica(replicaset)
                 i = i - 1
                 if r:is_connected() and (not prefer_replica or r ~= master) and
-                    replica_check_backoff(r, now)
-                then
-                    -- Pick a replica according prefered zone (highest priority replica zone) in round-robin manner
-                    if balance == "prefer_zone" and prefered_zone then
-                        local cbi = replicaset.balance_i
-                        local nr = replicaset_balance_replica(replicaset)
-
-                        if prefered_zone and r.zone and r.zone == prefered_zone
-                            and nr.zone and nr.zone ~= prefered_zone
-                            and (not prefer_replica or nr ~= master)
-                        then
-                            -- Reset cursor to the main position if next replica is in different zone.
-                            replicaset.balance_i = 1
-                        else
-                            -- Restore rr-cursor position.
-                            replicaset.balance_i = cbi
-                        end
-                    end
-
+                   replica_check_backoff(r, now) and
+                   (not prefered_zone or prefered_zone == r.zone) then
                     return r
                 end
             end
```
Of course, we would need tests to be sure.